3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,230 @@
+      return;
+    }
+
+    /**
+     * Multiple creatives on a page may wish to use the same type of
+     * amp-analytics tag. This object provides a mapping between the
+     * IDs which identify which amp-analytics tag a message is to/from,
+     * with each ID's corresponding AmpAnalytics3pCreativeMessageRouter,
+     * which is an object that handles messages to/from a particular creative.
+     * @private {!Object<string, !AmpAnalytics3pCreativeMessageRouter>}
+     */
+    this.creativeMessageRouters_ = {};
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(this.sentinel_);
+    this.iframeMessagingClient_.registerCallback(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE,
+        messageContainer => {
+          let entries;
+          dev().assert(messageContainer.data &&
+              (entries = Object.entries(messageContainer.data)).length,
+              'Received empty new creative message');
+          entries.forEach(entry => {
+            const creativeId = entry[0];
+            const extraData = entry[1];
+            dev().assert(!this.creativeMessageRouters_[creativeId],
+                'Duplicate new creative message for ' + creativeId);
+            this.creativeMessageRouters_[creativeId] =
+              new AmpAnalytics3pCreativeMessageRouter(
+                this.win_, this.iframeMessagingClient_, creativeId, extraData);
+          });
+        });
+    this.iframeMessagingClient_.registerCallback(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT,
+        messageContainer => {
+          let entries;
+          dev().assert(messageContainer.data &&
+              (entries = Object.entries(messageContainer.data)).length,
+              'Received empty events message');
+          entries.forEach(entry => {
+            const creativeId = entry[0];
+            const messages = entry[1];
+            try {
+              dev().assert(messages && messages.length,
+                  'Received empty events list for' + creativeId);
+              dev().assert(this.creativeMessageRouters_[creativeId],
+                  'Discarding event message received prior to new creative' +
+                  ' message for' + creativeId);
+              this.creativeMessageRouters_[creativeId]
+                  .sendMessagesToListener(messages);
+            } catch (e) {
+              dev().error(TAG_, 'Failed to pass message to event listener: ' +
+                  e.message);
+            }
+          });
+        });
+    this.iframeMessagingClient_.sendMessage(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.READY);
+  }
+}
+
+if (window.onNewAmpAnalyticsInstance) {
+  const router = new AmpAnalytics3pMessageRouter(window);
+  try {
+    window.onNewAmpAnalyticsInstance(router);
+  } catch (e) {
+    dev().error(TAG_, 'Exception thrown by onNewAmpAnalyticsInstance in' +
```

Comment:
I believe if you just include the error as another parameter AMP formats it correctly for reporting (e.g. dev().error(TAG_, 'Exception... ' + this.win_.location.href, e);)